test-data/unit/check-literal.test

```diff
@@ -0,0 +1,644 @@
+from typing_extensions import Literal as Foo
+
+x: Foo["Foo"]
+reveal_type(x)  # E: Revealed type is 'Literal['Foo']'
+
+y: Foo[Foo]     # E: Literal[...] must have at least one parameter
+[out]
+
+--
+-- Check to make sure we can construct the correct range of literal
+-- types (and correctly reject invalid literal types)
+--
+-- Note: the assignment tests exercise the logic in 'fastparse.py';
+-- the type alias tests exercise the logic in 'exprtotype.py'.
+--
+
+[case testLiteralBasicIntUsage]
+from typing_extensions import Literal
+
+a1: Literal[4]
+b1: Literal[0x2a]
+c1: Literal[-300]
+
+reveal_type(a1)  # E: Revealed type is 'Literal[4]'
+reveal_type(b1)  # E: Revealed type is 'Literal[42]'
+reveal_type(c1)  # E: Revealed type is 'Literal[-300]'
+
+a2t = Literal[4]
+b2t = Literal[0x2a]
+c2t = Literal[-300]
+a2: a2t
+b2: b2t
+c2: c2t
+
+reveal_type(a2)  # E: Revealed type is 'Literal[4]'
+reveal_type(b2)  # E: Revealed type is 'Literal[42]'
+reveal_type(c2)  # E: Revealed type is 'Literal[-300]'
+[out]
+
+[case testLiteralBasicBoolUsage]
+from typing_extensions import Literal
+
+a1: Literal[True]
+b1: Literal[False]
+
+reveal_type(a1)  # E: Revealed type is 'Literal[True]'
+reveal_type(b1)  # E: Revealed type is 'Literal[False]'
+
+a2t = Literal[True]
+b2t = Literal[False]
+a2: a2t
+b2: b2t
+
+reveal_type(a2)  # E: Revealed type is 'Literal[True]'
+reveal_type(b2)  # E: Revealed type is 'Literal[False]'
+[builtins fixtures/bool.pyi]
+[out]
+
+[case testLiteralBasicStrUsage]
+from typing_extensions import Literal
+
+a: Literal[""]
+b: Literal["  foo bar  "]
+c: Literal['  foo bar  ']
+
+reveal_type(a)  # E: Revealed type is 'Literal['']'
+reveal_type(b)  # E: Revealed type is 'Literal['  foo bar  ']'
+reveal_type(c)  # E: Revealed type is 'Literal['  foo bar  ']'
+[out]
+
+[case testLiteralBasicStrUsageSlashes]
+from typing_extensions import Literal
+
+a: Literal[r"foo\nbar"]
+b: Literal["foo\nbar"]
+
+reveal_type(a)
+reveal_type(b)
+[out skip-path-normalization]
+main:6: error: Revealed type is 'Literal['foo\\nbar']'
+main:7: error: Revealed type is 'Literal['foo\nbar']'
+
+[case testLiteralBasicNoneUsage]
+from typing_extensions import Literal
+a: Literal[None]
+reveal_type(a)   # E: Revealed type is 'None'
+# Note: Literal[None] and None are equivalent
+[out]
+
+[case testLiteralDisallowAny]
+from typing import Any
+from typing_extensions import Literal
+from missing_module import BadAlias     # E: Cannot find module named 'missing_module' \
+                                        # N: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
+
+a: Literal[Any]                         # E: Parameter 1 of Literal[...] is of type "Any"
+b: Literal[BadAlias]                    # E: Parameter 1 of Literal[...] is of type "Any"
+[out]
+
+[case testLiteralDisallowFloats]
+from typing_extensions import Literal
+a1: Literal[3.14]    # E: Parameter 1 of Literal[...] is of type "float"
```

Comment:
I prefer error messages that say what is violated, instead of just stating the fact. I mean error message `Parameter 1 of Literal[...] is of type "float"` doesn't really sound as an error. I would expect something like `Parameter 1 of Literal[...] cannot be of type "float"` or similar, for example `Parameter 1 of Literal[...] can be only of type "int", "str", "bool", or "Enum"`.